### ISSUE TITLE
Impossible to free GPU memory used by rank 0 tensors

### ISSUE THREAD

**Issue by Liedalyss** [tensorflow/tensorflow#92103]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

binary

### TensorFlow version

2.15.1

### Custom code

Yes

### OS platform and distribution

Ubuntu 22.04.5 LTS

### Python version

3.11.0rc1

### CUDA/cuDNN version

12.4

### Current behavior?

Whenever a tensor of rank 0 is created, it's impossible to be removed from GPU memory.
I tried also tf.keras.backend.clear_session() with no success.

The code I present is just one case. The same issues happens whenever a rank 0 tensor is created as a collateral (i.e. in tf.zeros,, when building the learning rate in an optimizer from a numerical value, or, probably,, when creating the variable optimizer._iterations)

### Standalone code to reproduce the issue

```shell
This is one example of code that creates the issue. I refer to it as code 1 in the relevant log output section:
import gc
import tensorflow as tf
# I create a rank 0 tensor
test=tf.cast(1.0,tf.float32)
del test
gc.collect()
# I force an oom to see the memory dump (any suggestion to a more graceful way to get the same output?)
crash = tf.ones([1000, 1000, 1000, 1000, 1000], tf.float32)


The following is the code I ran to check how the clean memory looks like (any suggestions for a cleaner way to get the same output?). I refer to this as code 2, in the relevant log output section:
import tensorflow as tf
# I force an oom to see the memory dump (any suggestion to a more graceful way to get the same output?)
crash = tf.ones([1000, 1000, 1000, 1000, 1000], tf.float32)
```

### Relevant log output

```shell
The following is the dump from code 1. Note the extra allocation for size 4, compared to the subsequent dump from code 2:
2025-04-24 03:26:32.437963: E external/local_xla/xla/stream_executor/gpu/gpu_cudamallocasync_allocator.cc:63] Histogram of current allocation: (allocation_size_in_bytes, nb_allocation_of_that_sizes), ...;
2025-04-24 03:26:32.437965: E external/local_xla/xla/stream_executor/gpu/gpu_cudamallocasync_allocator.cc:66] 4, 2
2025-04-24 03:26:32.437967: E external/local_xla/xla/stream_executor/gpu/gpu_cudamallocasync_allocator.cc:66] 1028, 1
2025-04-24 03:26:32.437970: E external/local_xla/xla/stream_executor/gpu/gpu_cudamallocasync_allocator.cc:97] CU_MEMPOOL_ATTR_RESERVED_MEM_CURRENT: 33554432



The following is a clean tensorflow dump, from code 2:
2025-04-24 03:23:41.062128: E external/local_xla/xla/stream_executor/gpu/gpu_cudamallocasync_allocator.cc:66] 4, 1
2025-04-24 03:23:41.062129: E external/local_xla/xla/stream_executor/gpu/gpu_cudamallocasync_allocator.cc:66] 1028, 1
2025-04-24 03:23:41.062132: E external/local_xla/xla/stream_executor/gpu/gpu_cudamallocasync_allocator.cc:97] CU_MEMPOOL_ATTR_RESERVED_MEM_CURRENT: 33554432
```

**Comment by KAVYANSHTYAGI** [comment]
I looked into the issue and can confirm the behavior—rank-0 tensors like tf.cast(1.0, tf.float32) seem to persist on GPU memory even after del and gc.collect(). I reproduced the behavior using TensorFlow 2.15.1 on Ubuntu 22.04 with CUDA 12.4. The extra 4-byte allocation remains unless the entire session is reset.

A few things I tried:

tf.keras.backend.clear_session() → no effect

gc.collect() → no effect

Setting memory growth with:

gpus = tf.config.experimental.list_physical_devices('GPU')
for gpu in gpus:
    tf.config.experimental.set_memory_growth(gpu, True)

seems to reduce overall pooling behavior, but the rank-0 tensor allocation still lingers

From the logs, it looks like this could be due to the cudaMallocAsync allocator caching small allocations for reuse. Would it be possible to allow opt-out of caching for small tensors like scalars? Or expose an API to explicitly release them from memory?

Happy to test any fix or workaround. Let me know if further logs or traces would help!

**Comment by Liedalyss** [comment]
Thank you for checking  :)
Maybe it's worth noting that rank 1 float 32 tensors with just one element are correctly removed from GPU memory.

**Comment by Venkat6871** [comment]
Hi @Liedalyss ,
Apologies for the delay, and thank you for raising your concern here. I tried running your code on Colab using TensorFlow version 2.19.0 and did not encounter the issue you described. Could you please check on your end? For your reference, I am attaching a [gist](https://colab.sandbox.google.com/gist/Venkat6871/18a67aa007ca3910e98dea49666682ca/92103_tf_2-19-0-v-gpu.ipynb). It seems that a version compatibility mismatch might be causing the issue. Could you please verify that all components are compatible for better results? Additionally, I noticed that you are using an older version please update to the latest version for improved stability and support. Here, I am providing the official [documentation](https://www.tensorflow.org/install/source#gpu) for your reference. If you are still facing any issues, please let us know we are happy to assist further.
Thank you!

**Comment by Liedalyss** [comment]
Hi @Venkat6871 

sorry for the delay.

In your gist I don't see any memory usage printout like the one I attached: could it by that it's not shown at all in the colab?

I need no maintain the possibility to use the current version till some calls to tensorflow addons will be phased out, but trying the latest tensorflow version would require updating CUDA and cuDNN according to your chart.
Do you have any official guide to allow for multiple versions of CUDA and cuDNN to coexist on the same machine?
Thank you.

**Comment by Kayyuri** [comment]
Hi @Liedalyss,
Thank you so much for filing this issue! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

Thank you again for your interest and we look forward to seeing your contributions!

**Comment by Kayyuri** [comment]
Hi @Liedalyss,
We are closing this issue due to inactivity. If this is still relevant and you would like to work on it, please feel free to reopen the issue or submit a pull request referencing it.

Thank you for your understanding!
